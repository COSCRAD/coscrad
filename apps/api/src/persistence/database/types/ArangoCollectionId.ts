/**
 * TODO [https://www.pivotaltracker.com/story/show/182132515]
 * Refactor. Reorganize our collection name \ type management.
 */
const arangoResourceCollectionIDs = [
    'terms',
    'vocabulary_lists',
    'transcribed_audio',
    'books',
    'photographs',
    'spatial_features',
] as const;

export const categoryEdgeCollectionID = 'category_edges';

export const edgeConnectionCollectionID = 'resource_edge_connections';

export const arangoEdgeCollectionIDs = [categoryEdgeCollectionID, edgeConnectionCollectionID];

export const isArangoEdgeCollectionCollectionID = (input: unknown): boolean =>
    arangoEdgeCollectionIDs.includes(input as string);

export const tagCollectionID = 'tags';

export const categoryCollectionID = 'categories';

const arangoCollectionIDs = [
    ...arangoResourceCollectionIDs,
    ...arangoEdgeCollectionIDs,
    tagCollectionID,
    categoryCollectionID,
] as const;

export type ArangoResourceCollectionID = typeof arangoResourceCollectionIDs[number];

export const isArangoResourceCollectionID = (input: unknown): input is ArangoResourceCollectionID =>
    arangoResourceCollectionIDs.includes(input as ArangoResourceCollectionID);

export const getAllArangoResourceCollectionIDs = (): ArangoResourceCollectionID[] => [
    ...arangoResourceCollectionIDs,
];

export type ArangoCollectionID = typeof arangoCollectionIDs[number];

export const getAllArangoCollectionIDs = (): ArangoCollectionID[] => [...arangoCollectionIDs];
