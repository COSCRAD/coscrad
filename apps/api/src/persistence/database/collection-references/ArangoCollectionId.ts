import { ValueType } from '../../../lib/types/valueType';
import { ArangoDocumentCollectionId } from './ArangoDocumentCollectionId';
import { ArangoEdgeCollectionId } from './ArangoEdgeCollectionId';

export const ArangoCollectionId = {
    ...ArangoDocumentCollectionId,
    ...ArangoEdgeCollectionId,
};

/**
 * We are stacking this type with the above value definition because there doesn't
 * seem to be a more natural way to compose enums in TypeScript at present.
 */
export type ArangoCollectionId = ValueType<typeof ArangoCollectionId>;

export const getAllArangoCollectionIDs = () => Object.values(ArangoCollectionId);

export const isArangoCollectionId = (input: unknown): input is ArangoCollectionId =>
    getAllArangoCollectionIDs().some((id) => id === input);
