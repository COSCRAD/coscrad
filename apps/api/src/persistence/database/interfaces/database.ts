import { ResourceId } from 'apps/api/src/domain/types/ResourceId';
import { Maybe } from 'apps/api/src/lib/types/maybe';

export interface IDatabase {
    fetchById: <TEntityDTO>(id: ResourceId, collectionName: string) => Promise<Maybe<TEntityDTO>>;

    fetchMany: <TEntityDTO>(collectionName: string) => Promise<Maybe<TEntityDTO[]>>;

    create: <TEntityDTO>(dto: TEntityDTO, collectionName: string) => Promise<void>;

    createMany: <TEntityDTO>(dto: TEntityDTO[], collectionName: string) => Promise<void>;

    getCount: (collectionName: string) => Promise<number>;

    update: <TEntityDTO>(id: ResourceId, dto: TEntityDTO, collectionName: string) => Promise<void>;
}
