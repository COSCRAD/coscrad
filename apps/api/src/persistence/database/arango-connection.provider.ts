import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { Database } from 'arangojs';
import { Scheme } from '../../app/config/constants/Scheme';
import { isNullOrUndefined } from '../../domain/utilities/validation/is-null-or-undefined';
import { DTO } from '../../types/DTO';
import ArangoDatabaseConfiguration from './ArangoDatabaseConfiguration';
import { getAllArangoCollectionIDs } from './collection-references/ArangoCollectionId';
import { isArangoEdgeCollectionCollectionID } from './collection-references/ArangoEdgeCollectionId';
import DatabaseAlreadyInitializedError from './errors/DatabaseAlreadyInitializedError';
import DatabaseCannotBeDroppedError from './errors/DatabaseCannotBeDroppedError';
import DatabaseNotYetInitializedError from './errors/DatabaseNotYetInitializedError';
import canDatabaseBeDropped from './utilities/canDatabaseBeDropped';

// Alias for more clarity from the outside; TODO wrap `Database` with simpler API?
export type ArangoConnection = Database;

type Port = `${number}`;

const isPortRequired = (scheme: Scheme, port: `${number}`) => {
    if (scheme === 'http' && port === '80') return false;
    if (scheme === 'https' && port === '443') return false;
    return true;
};

const buildFullHostURL = (
    domain: string,
    scheme: Scheme = Scheme.https,
    port: Port = '443'
): string => `${scheme}://${domain}${isPortRequired(scheme, port) ? `:${port}` : ''}`;

type AdminDatabaseConnectionOptions = {
    shouldConnectToSystemDatabase: boolean;
};

@Injectable()
export class ArangoConnectionProvider {
    #connection: ArangoConnection;

    #databaseConfiguration: ArangoDatabaseConfiguration;

    isInitialized = false;

    public get databaseConfiguration() {
        return this.#databaseConfiguration.getConfig();
    }

    constructor(private configService: ConfigService) {
        this.setDatabaseConfiguration({
            dbName: this.configService.get<string>('ARANGO_DB_NAME'),
            dbRootPass: this.configService.get<string>('ARANGO_DB_ROOT_PASSWORD'),
            dbUser: this.configService.get<string>('ARANGO_DB_USER'),
            dbPass: this.configService.get<string>('ARANGO_DB_USER_PASSWORD'),
            dbHostUrl: buildFullHostURL(
                this.configService.get<string>('ARANGO_DB_HOST_DOMAIN', 'localhost'),
                this.configService.get<Scheme>('ARANGO_DB_HOST_SCHEME', Scheme.http),
                this.configService.get<number>('ARANGO_DB_HOST_PORT', 80).toString() as Port
            ),
        });

        console.log(`dbName: ${this.databaseConfiguration.dbName}, dbRootPass: ${this.databaseConfiguration.dbRootPass}, dbUser: ${this.databaseConfiguration.dbUser}, dbPass: ${this.databaseConfiguration.dbPass}, dbHostUrl: ${this.databaseConfiguration.dbHostUrl}`);
        

        const systemDB = this.#getAdminDBConnection({
            shouldConnectToSystemDatabase: true,
        });

        // TODO why is `useDatabase` deprecated? can we use myDB.database("db_name")?
        const dbInstance = systemDB.database(this.databaseConfiguration.dbName);
        dbInstance.useBasicAuth(
            this.databaseConfiguration.dbUser,
            this.databaseConfiguration.dbPass
        );

        this.#connection = dbInstance;
    }

    /**
     * This async setup helper will
     * - Create the database (the one referenced in your .env via ConfigService)
     * only if this database does not exist
     * - Add any missing collections (but leave existing collections untouched)
     * and can only be run once.
     */
    async initialize(): Promise<void> {
        if (this.isInitialized) throw new DatabaseAlreadyInitializedError();

        await this.#createDatabaseIfNotExists();

        await this.#createAllMissingCollections();

        this.isInitialized = true;
    }

    getDatabaseName(): string {
        return this.databaseConfiguration.dbName;
    }

    getConnection(): ArangoConnection {
        if (!this.isInitialized)
            throw new DatabaseNotYetInitializedError('get an Arango connection');

        return this.#connection;
    }

    public setDatabaseConfiguration(config: DTO<ArangoDatabaseConfiguration>) {
        this.#databaseConfiguration = new ArangoDatabaseConfiguration({
            ...config,
        });
    }

    #getAdminDBConnection({
        shouldConnectToSystemDatabase,
    }: AdminDatabaseConnectionOptions): Database {
        const adminInstance = new Database({
            url: this.databaseConfiguration.dbHostUrl,
        });

        adminInstance.useBasicAuth('root', this.databaseConfiguration.dbRootPass);

        if (shouldConnectToSystemDatabase) {
            return adminInstance;
        }

        // TODO why is `useDatabase` deprecated? can we use myDB.database("db_name")?
        const dbInstance = adminInstance.database(this.databaseConfiguration.dbName);

        return dbInstance;
    }

    async #doesCollectionExist(collectionName: string): Promise<boolean> {
        const allCollections = await this.#connection
            .listCollections()
            .then((allCollectionMetadata) => allCollectionMetadata.map(({ name }) => name));

        const doesCollectionExist = allCollections.includes(collectionName);

        return doesCollectionExist;
    }

    async #createDatabaseIfNotExists(): Promise<void> {
        const databaseName = this.databaseConfiguration.dbName;

        if (await this.#doesDatabaseExist(databaseName)) return;

        const { dbUser } = this.#databaseConfiguration;

        await this.#getAdminDBConnection({
            shouldConnectToSystemDatabase: true,
        }).createDatabase(databaseName, {
            users: [{ username: dbUser }, { username: 'root' }],
        });
    }

    async #createAllMissingCollections(): Promise<void> {
        await Promise.all(
            getAllArangoCollectionIDs().map((collectionName) =>
                this.#createCollectionIfNotExists(collectionName)
            )
        );
    }

    async #createCollectionIfNotExists(collectionName: string): Promise<void> {
        const doesCollectionExist = await this.#doesCollectionExist(collectionName);

        if (doesCollectionExist) return;

        /**
         * TODO [https://www.pivotaltracker.com/story/show/182132515]
         * cleanup the references
         */
        if (isArangoEdgeCollectionCollectionID(collectionName)) {
            await this.#connection.createEdgeCollection(collectionName);
        } else {
            await this.#connection.createCollection(collectionName);
        }
    }

    #doesDatabaseExist = async (databaseName: string) => {
        if (isNullOrUndefined(databaseName)) return false;

        const doesDatabaseExist = await this.#getAdminDBConnection({
            shouldConnectToSystemDatabase: true,
        })
            .listDatabases()
            .then((allDatabaseNames) => allDatabaseNames.includes(databaseName));

        return doesDatabaseExist;
    };

    /**
     * **CAUTION** This method is only to be used in tests.
     */
    async dropDatabaseIfExists(): Promise<void> {
        const dbName = this.databaseConfiguration.dbName;

        const doesDatabaseExist = await this.#doesDatabaseExist(dbName);

        if (!doesDatabaseExist) return;

        // for safety, you can only drop a db whose name matches the test db name pattern
        if (!canDatabaseBeDropped(dbName)) throw new DatabaseCannotBeDroppedError(dbName);

        await this.#getAdminDBConnection({
            shouldConnectToSystemDatabase: true,
        }).dropDatabase(dbName);
    }
}
