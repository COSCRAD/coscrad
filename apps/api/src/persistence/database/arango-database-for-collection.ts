import { isArangoDatabase } from 'arangojs/database';
import { ISpecification } from '../../domain/repositories/interfaces/ISpecification';
import { AggregateId } from '../../domain/types/AggregateId';
import { HasAggregateId } from '../../domain/types/HasAggregateId';
import { Maybe } from '../../lib/types/maybe';
import { DeepPartial } from '../../types/DeepPartial';
import { ArangoDatabase } from './arango-database';
import { ArangoCollectionId } from './collection-references/ArangoCollectionId';
import { DatabaseDocument } from './utilities/mapEntityDTOToDatabaseDTO';

/**
 * Note that at this level we are working with a `DatabaseDocument` (has _key
 * and _id), not an `EntityDTO`. The mapping is taken care of in the
 * repositories layer.
 */
export class ArangoDatabaseForCollection<TEntity extends HasAggregateId> {
    #collectionID: ArangoCollectionId;

    #arangoDatabase: ArangoDatabase;

    constructor(arangoDatabase: ArangoDatabase, collectionName: ArangoCollectionId) {
        this.#collectionID = collectionName;

        this.#arangoDatabase = arangoDatabase;

        if (isArangoDatabase(this.#arangoDatabase))
            throw new Error(
                `Received invalid arango db instance: ${JSON.stringify(arangoDatabase)}`
            );
    }

    // Queries (return information)
    fetchById(id: AggregateId): Promise<Maybe<DatabaseDocument<TEntity>>> {
        return this.#arangoDatabase.fetchById<DatabaseDocument<TEntity>>(id, this.#collectionID);
    }

    fetchMany(specification?: ISpecification<TEntity>): Promise<DatabaseDocument<TEntity>[]> {
        return this.#arangoDatabase.fetchMany<DatabaseDocument<TEntity>>(
            this.#collectionID,
            // TODO remove cast, handle mapping layer
            specification as unknown as ISpecification<DatabaseDocument<TEntity>>
        );
    }

    getCount(): Promise<number> {
        return this.#arangoDatabase.getCount(this.#collectionID);
    }

    // Commands (mutate state)
    create(DatabaseDocument: DatabaseDocument<TEntity>) {
        // Handle the difference in _id \ _key between model and database
        return this.#arangoDatabase.create(DatabaseDocument, this.#collectionID);
    }

    createMany(DatabaseDocuments: DatabaseDocument<TEntity>[]) {
        return this.#arangoDatabase.createMany(DatabaseDocuments, this.#collectionID);
    }

    update(id: AggregateId, updateDTO: DeepPartial<DatabaseDocument<TEntity>>) {
        return this.#arangoDatabase.update(id, updateDTO, this.#collectionID);
    }
}
