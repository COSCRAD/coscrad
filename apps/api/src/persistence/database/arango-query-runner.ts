import { HasId } from '@coscrad/api-interfaces';
import { Injectable } from '@nestjs/common';
import { isDeepStrictEqual } from 'util';
import { ICoscradQueryRunner } from '../migrations/coscrad-query-runner.interface';
import { ArangoDatabase } from './arango-database';
import { ArangoCollectionId } from './collection-references/ArangoCollectionId';
import { ArangoDatabaseProvider } from './database.provider';
import { ArangoDatabaseDocument } from './utilities/mapEntityDTOToDatabaseDTO';

@Injectable()
export class ArangoQueryRunner implements ICoscradQueryRunner {
    private readonly arangoDatabase: ArangoDatabase;

    constructor(arangoDatabaseProvider: ArangoDatabaseProvider) {
        this.arangoDatabase = arangoDatabaseProvider.getDBInstance();
    }

    async update<TOldDocument extends ArangoDatabaseDocument<HasId>, UNewDocument>(
        collectionName: ArangoCollectionId,
        calculateUpdate: (oldDocument: TOldDocument) => UNewDocument
    ): Promise<void> {
        console.log(`running update queries with: ${this.arangoDatabase.getDatabaseName()}`);

        const existingDocs = await this.arangoDatabase.fetchMany<TOldDocument>(collectionName);

        const updates = existingDocs
            .map((oldDocument) => ({
                ...calculateUpdate(oldDocument),
                // this ensures that the update can be matched to an existing doc
                _key: oldDocument._key,
            }))
            // avoid unnecessary writes
            .filter((update) => !isDeepStrictEqual(update, {}));

        await this.arangoDatabase.updateMany(updates, collectionName);
    }
}
