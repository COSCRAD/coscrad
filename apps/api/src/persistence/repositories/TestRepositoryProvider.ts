import { Entity } from '../../domain/models/entity';
import { EntityType, InMemorySnapshot } from '../../domain/types/entityType';
import { DatabaseProvider } from '../database/database.provider';
import { getArangoCollectionIDFromEntityType } from '../database/get-arango-collection-ids';
import { RepositoryProvider } from './repository.provider';

export default class TestRepositoryProvider extends RepositoryProvider {
  constructor(databaseProvider: DatabaseProvider) {
    super(databaseProvider);
  }

  // TODO We should correlate entity type with TEntity here
  public async addEntitiesOfSingleType<TEntity extends Entity>(
    entityType: EntityType,
    entities: TEntity[]
  ): Promise<void> {
    await this.forEntity<TEntity>(entityType).createMany(entities);
  }

  // TODO fix types
  public async addEntitiesOfManyTypes(
    snapshot: InMemorySnapshot
  ): Promise<void> {
    const writePromises = Object.entries(snapshot).map(
      ([entityType, entityInstances]) =>
        this.addEntitiesOfSingleType(
          entityType as EntityType,
          entityInstances as any
        )
    );
  }

  public async deleteAllEntitiesOfGivenType(
    entityType: EntityType
  ): Promise<void> {
    await (
      await this.databaseProvider.getDBInstance()
    ).deleteAll(getArangoCollectionIDFromEntityType(entityType));
  }
}
