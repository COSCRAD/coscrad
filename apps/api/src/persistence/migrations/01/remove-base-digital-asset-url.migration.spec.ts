import { isDeepStrictEqual } from 'util';
import createTestModule from '../../../app/controllers/__tests__/createTestModule';
import { CoscradEventFactory } from '../../../domain/common';
import buildDummyUuid from '../../../domain/models/__tests__/utilities/buildDummyUuid';
import { Photograph } from '../../../domain/models/photograph/entities/photograph.entity';
import { Term } from '../../../domain/models/term/entities/term.entity';
import { ResourceType } from '../../../domain/types/ResourceType';
import { isNullOrUndefined } from '../../../domain/utilities/validation/is-null-or-undefined';
import { InternalError } from '../../../lib/errors/InternalError';
import cloneToPlainObject from '../../../lib/utilities/cloneToPlainObject';
import { DTO } from '../../../types/DTO';
import { DynamicDataTypeFinderService } from '../../../validation';
import { ArangoConnectionProvider } from '../../database/arango-connection.provider';
import { ArangoQueryRunner } from '../../database/arango-query-runner';
import { ArangoCollectionId } from '../../database/collection-references/ArangoCollectionId';
import { ArangoDatabaseProvider } from '../../database/database.provider';
import {
    ArangoDatabaseDocument,
    ArangoDocumentForAggregateRoot,
} from '../../database/utilities/mapEntityDTOToDatabaseDocument';
import TestRepositoryProvider from '../../repositories/__tests__/TestRepositoryProvider';
import generateDatabaseNameForTestSuite from '../../repositories/__tests__/generateDatabaseNameForTestSuite';
import {
    BASE_DIGITAL_ASSET_URL,
    RemoveBaseDigitalAssetUrl,
} from './remove-base-digital-asset-url.migration';

const baseDigitalAssetUrl = `https://www.mymedia.org/downloads/`;

process.env[BASE_DIGITAL_ASSET_URL] = baseDigitalAssetUrl;

type OldPhotograph = Omit<Photograph, 'imageUrl'> & { filename?: string };

/**
 * Migrations are a vestige of when we used a state-based approach to persisting
 * the domain state.
 *
 * This migration has been applied to all databases. We keep this test for
 * posterity.
 */
describe.skip(`RemoveBaseDigitalAssetUrl`, () => {
    let testDatabaseProvider: ArangoDatabaseProvider;

    let testQueryRunner: ArangoQueryRunner;

    let testRepositoryProvider: TestRepositoryProvider;

    const buildId = (resourceType: ResourceType, index: number): string => {
        if (resourceType === ResourceType.term) return buildDummyUuid(index);

        if (resourceType === ResourceType.photograph) return buildDummyUuid(100 + index);

        throw new InternalError(
            `resource type: ${resourceType} is not supported in this ID generation scheme`
        );
    };

    describe(`when there are documents (terms, vocabularyLists, and photographs) that should be updated`, () => {
        beforeAll(async () => {
            const testModule = await createTestModule({
                ARANGO_DB_NAME: generateDatabaseNameForTestSuite(),
            });

            const arangoConnectionProvider = testModule.get(ArangoConnectionProvider);

            testDatabaseProvider = new ArangoDatabaseProvider(arangoConnectionProvider);

            const coscradEventFactory = testModule.get(CoscradEventFactory);

            const databaseProvider = testModule.get(ArangoDatabaseProvider);

            /**
             * It's a bit awkward that we need this because we are not working at
             * the repositories level of abstraction. However, we have added test
             * setup and teardown logic at this level for the purpose of command and
             * query integration tests. So instead of rewriting this logic on a
             * `TestDatabaseProvider`, we will just leverage this existing logic for
             * test teardown.
             */
            testRepositoryProvider = new TestRepositoryProvider(
                databaseProvider,
                coscradEventFactory,
                testModule.get(DynamicDataTypeFinderService)
            );

            testQueryRunner = new ArangoQueryRunner(testDatabaseProvider);
        });

        // TERMS
        const dtoForTermToCheckManually = {
            term: `so bogus`,
            termEnglish: `so bogus (English)`,
            audioFilename: `bogus`,
            type: ResourceType.term,
            published: true,
            contributorId: '55',
        };

        const dtoForTermWithoutAudioFilename = {
            term: `so bogus`,
            termEnglish: `so bogus (English)`,
            // audioFilename: MISSING!,
            type: ResourceType.term,
            published: true,
            contributorId: '55',
        };

        const originalTermDocumentsWithoutKeys = [
            dtoForTermToCheckManually,
            dtoForTermWithoutAudioFilename,
        ];

        const originalTermDocuments = originalTermDocumentsWithoutKeys.map((partialDto, index) => ({
            ...partialDto,
            _key: buildId(ResourceType.term, index),
        }));

        // PHOTOGRAPHS
        const dtoForPhotographToCheckManually = {
            type: ResourceType.photograph,
            filename: `flowers`,
            photographer: `James Rames`,
            dimensions: {
                widthPX: 300,
                heightPX: 400,
            },
            published: true,
        };

        const dtoForPhotographWithoutFilename = {
            type: ResourceType.photograph,
            // filename: MISSING!,
            photographer: `James Rames`,
            dimensions: {
                widthPX: 300,
                heightPX: 400,
            },
            published: true,
        };

        const originalPhotographDocumentsWithoutKeys = [
            dtoForPhotographToCheckManually,
            dtoForPhotographWithoutFilename,
        ];

        const originalPhotographDocuments = originalPhotographDocumentsWithoutKeys.map(
            (partialDto, index) => ({
                ...partialDto,
                _key: buildId(ResourceType.photograph, index),
            })
        );

        beforeEach(async () => {
            // We use the `TestRepositoryProvider` for test teardown only
            await testRepositoryProvider.testTeardown();

            await testDatabaseProvider
                .getDatabaseForCollection(ArangoCollectionId.terms)
                .createMany(originalTermDocuments);

            await testDatabaseProvider
                .getDatabaseForCollection(ArangoCollectionId.photographs)
                .createMany(originalPhotographDocuments);
        });

        const migrationUnderTest = new RemoveBaseDigitalAssetUrl();

        const idForTermToCheckManually = buildId(ResourceType.term, 0);

        const idForTermWithoutAudioFilename = buildId(ResourceType.term, 1);

        const idForPhotographToCheckManually = buildId(ResourceType.photograph, 0);

        const idForPhotographWithoutFilename = buildId(ResourceType.photograph, 1);

        it(`should apply the appropriate updates`, async () => {
            await migrationUnderTest.up(testQueryRunner);

            const updatedTermDocuments = (await testDatabaseProvider
                .getDatabaseForCollection(ArangoCollectionId.terms)
                .fetchMany()) as ArangoDocumentForAggregateRoot<DTO<Term>>[];

            expect(updatedTermDocuments.length).toBe(originalTermDocuments.length);

            const idsOfDocumentsWithoutBaseDigitalAssetUrl = updatedTermDocuments.filter(
                // @ts-expect-error no longer relevant
                ({ audioItemId: audioFilename }) =>
                    !isNullOrUndefined(audioFilename) &&
                    !audioFilename.includes(baseDigitalAssetUrl)
            );

            expect(idsOfDocumentsWithoutBaseDigitalAssetUrl).toEqual([]);

            // @ts-expect-error this test is not longer relevant
            const { audioItemId: audioFilename } = (await testDatabaseProvider
                .getDatabaseForCollection(ArangoCollectionId.terms)
                .fetchById(idForTermToCheckManually)) as unknown as ArangoDatabaseDocument<
                DTO<Term>
            >;

            expect(audioFilename).toBe(`https://www.mymedia.org/downloads/bogus.mp3`);

            const migratedTermWithoutAudiofilename = (await testDatabaseProvider
                .getDatabaseForCollection(ArangoCollectionId.terms)
                .fetchById(idForTermWithoutAudioFilename)) as unknown as ArangoDatabaseDocument<
                DTO<Term>
            >;

            // @ts-expect-error this test is no longer relevant
            expect(migratedTermWithoutAudiofilename.audioItemId).not.toBeTruthy();

            // @ts-expect-error there's no point of maintainging this any longer
            const { imageUrl } = (await testDatabaseProvider
                .getDatabaseForCollection(ArangoCollectionId.photographs)
                .fetchById(idForPhotographToCheckManually)) as unknown as ArangoDatabaseDocument<
                DTO<Photograph>
            >;

            expect(imageUrl).toBe(`https://www.mymedia.org/downloads/flowers.png`);

            const dtoForPhotographWithoutFilename = (await testDatabaseProvider
                .getDatabaseForCollection(ArangoCollectionId.photographs)
                .fetchById(idForPhotographWithoutFilename)) as unknown as ArangoDatabaseDocument<
                DTO<Photograph>
            >;

            // @ts-expect-error There's no reason to support this now
            expect(dtoForPhotographWithoutFilename.imageUrl).not.toBeTruthy();

            const updatedPhotographDocuments = (await testDatabaseProvider
                .getDatabaseForCollection(ArangoCollectionId.photographs)
                .fetchMany()) as ArangoDocumentForAggregateRoot<DTO<Photograph>>[];

            expect(updatedPhotographDocuments.length).toBe(originalPhotographDocuments.length);

            const idsOfPhotographDocumentsWithoutBaseDigitalAssetUrl =
                updatedPhotographDocuments.filter(
                    // @ts-expect-error There's no reason to support this now
                    ({ imageUrl }) =>
                        !isNullOrUndefined(imageUrl) && !imageUrl.includes(baseDigitalAssetUrl)
                );

            expect(idsOfPhotographDocumentsWithoutBaseDigitalAssetUrl).toEqual([]);

            const idsOfPhotographsWithVestigialFilenameProperty = updatedPhotographDocuments.filter(
                (newDoc) => !isNullOrUndefined((newDoc as unknown as OldPhotograph).filename)
            );

            expect(idsOfPhotographsWithVestigialFilenameProperty).toEqual([]);
        });

        it(`should be reverseable`, async () => {
            // TODO Consider adding a custom jest matcher instead
            const removeRevFromDoc = (document: { _rev: string }) => {
                const cloned = cloneToPlainObject(document);

                delete cloned['_rev'];

                return cloned;
            };

            const buildMiniSnapshot = async () => {
                const photographs = await testDatabaseProvider
                    .getDatabaseForCollection(ArangoCollectionId.photographs)
                    .fetchMany();

                const terms = await testDatabaseProvider
                    .getDatabaseForCollection(ArangoCollectionId.terms)
                    .fetchMany();

                return {
                    [ArangoCollectionId.photographs]: photographs.map((photograph) =>
                        removeRevFromDoc(photograph as unknown as { _rev: string })
                    ),
                    [ArangoCollectionId.terms]: terms.map((term) =>
                        removeRevFromDoc(term as unknown as { _rev: string })
                    ),
                };
            };

            const { terms: termsBefore, photographs: photographsBefore } =
                await buildMiniSnapshot();

            await migrationUnderTest.up(testQueryRunner);

            await migrationUnderTest.down(testQueryRunner);

            const { terms: termsAfter, photographs: photographsAfter } = await buildMiniSnapshot();

            const doArraysContainEqualMembers = <T, U>(a: T[], b: U[]) =>
                a.length === b.length &&
                a.every((element) =>
                    b.some((elementFromB) => isDeepStrictEqual(element, elementFromB))
                );

            // TODO add a custom jest matcher for this
            // TODO can we optimize this?
            expect(doArraysContainEqualMembers(termsBefore, termsAfter)).toBe(true);

            expect(doArraysContainEqualMembers(photographsBefore, photographsAfter)).toBe(true);
        });
    });
});
