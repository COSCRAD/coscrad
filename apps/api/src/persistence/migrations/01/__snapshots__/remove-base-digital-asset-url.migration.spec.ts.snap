// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoveBaseDigitalAssetUrl when there are documents (terms, vocabularyLists, and photographs) that should be updated should apply the appropriate updates 1`] = `Promise {}`;
